### PR TITLE
#116 Relative stability

### PR DESCRIPTION
**Pull request by aaarrti**
### Description
- This PR implements new metrics requested in this [#116](https://github.com/understandable-machine-intelligence-lab/Quantus/issues/116) issue
- Additionally, there are some refactorings, I just could not ignore
### Implemented changes
- Actual metric implementation.
  - [x] Added `quantus.metrics.robustness.RelativeInputStability`
  - [x] Added `quantus.metrics.robustness.RelativeOutputStability`
  - [x] Added `quantus.metrics.robustness.RelativeRepresentationStability`
  - [x] Updated `Tutorial_ImageNet_Example_All_Metrics.ipynb` with these metrics
  - [x] Updated `Tutorial_Getting_Started_with_Tensorflow.ipynb` with these metrics
  - [x] Added test cases for these metrics
  - [x] Generated, `.rst` files for newly added classes.
  - [x] Updated README
  - [x] Added new method `get_hidden_representations` to `quantus.helpers.models.ModelInterface`, 
  as it is required for  `RelativeRepresentationStability`. Obviously, also added its implementation in TensorFlow and PyTorch.
  - [x] Added `generate_normalized_explanations_batch` to `BatchedPerturbationMetric`, all 3 Relative Stability metrics required the logic inside. Probably, this could also be used by other metrics.

- Other changes
  - [x] Added step to `.github/workflows/python-package.yml`, which install library locally, and tests if import doesn't fail.
  - [x] `TFLeNet` was not a LeNet architecture, updated it, trained it, added weights.
  - [x] Added 2D CNN implementation in TensorFlow because there was not TF model, which supported RGB images. Also trained it, and added weights.
  - [x] Added `random_uniform_noise` perturb function.
  - [x] Replaced assertion against real model outputs, with assertion against mock.
  - [x] Performance improvement for TFModel. Now instead of rebuilding models on each `.predict()` and `.get_hidden_representations()` call, the rebuilt models are cached.
  - [x] Removed conversion `np.ndarray` -> `torch.Tensor` -> `np.ndarray` in `fixtures.py`. Now, data is just loaded as `np.ndarray`s

### CODE REVIEW

File: .github/workflows/python-package.yml [hunk 1]
```diff
@@ -44,3 +44,8 @@ jobs:
     - name: Test with pytest
       run: |
         pytest
+
+    - name: Test importing the library
+      run: |
+        pip install .
```
Comment (annahedstroem) [inline .github/workflows/python-package.yml:50]: Great thank you!

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -68,6 +68,13 @@ To perform the tests, execute the following (make sure pytest is installed):
 ```bash
 pytest
 ```
+
+
```
Comment (annahedstroem) [inline CONTRIBUTING.md:72]: super

File: quantus/functions/perturb_func.py [hunk 3]
```diff
@@ -556,3 +555,31 @@ def no_perturbation(arr: np.array, **kwargs) -> np.array:
...
+
+
+def random_uniform_noise(arr: np.ndarray, amplitude=1e-6, **kwargs) -> np.ndarray:
+    """
+    Parameters
+    ----------
+    arr: np.ndarray
+         3D tensor representing image
+    amplitude: float
+                 Upper bound for noise generated, default 1e-6
```
Comment (annahedstroem) [inline quantus/functions/perturb_func.py:567]: too much indent

File: quantus/functions/perturb_func.py [hunk 4]
```diff
@@ -556,3 +555,31 @@ def no_perturbation(arr: np.array, **kwargs) -> np.array:
...
+    """
+    arr_shape = arr.shape
+    if len(arr_shape) == 4:
+        noise_shape = arr.shape[1:]
+    elif len(arr_shape) == 3:
+        noise_shape = arr_shape
+    else:
+        raise ValueError("random_noise supports only 3D and 4D")
+
+    noise = np.random.uniform(size=noise_shape, low=-1 * amplitude, high=amplitude)
```
Comment (annahedstroem) [inline quantus/functions/perturb_func.py:584]: why not `low=-amplitude`?

File: quantus/helpers/model/model_interface.py [hunk 5]
```diff
@@ -105,3 +105,36 @@ def get_random_layer_generator(self):
...
+        should be, we leave it up to user whether all layers are used,
+        or couple specific ones should be selected.
+        User can select layer by providing 'layer_names' (exclusive)OR 'layer_indices'.
+
+        Parameters
+        ----------
+        x: np.ndarray
+           4D tensor, a batch of input datapoints
+        layer_names: List[str]
+                     List with names of layers, from which output should be captured.
```
Comment (annahedstroem) [inline quantus/helpers/model/model_interface.py:130]: indents

File: quantus/helpers/model/model_interface.py [hunk 6]
```diff
@@ -105,3 +105,36 @@ def get_random_layer_generator(self):
...
+
+    @abstractmethod
+    def get_hidden_representations(
+        self,
+        x: np.ndarray,
+        layer_names: Optional[List[str]] = None,
+        layer_indices: Optional[List[int]] = None,
+    ) -> np.ndarray:
+        """
+        Computes models internal representation of input x.
```
Comment (annahedstroem) [inline quantus/helpers/model/model_interface.py:117]: Rewrite to:

Compute the model's internal representation of input x.
In practice, this means, executing a forward pass and then, capturing the output of layers (of interest).
As the exact definition of  "internal model representation" is left out in the original paper (see:  https://arxiv.org/pdf/2203.06877.pdf), we make the implementation flexible.
It is up to the user whether all layers are used, or specific ones should be selected.
The user can therefore select a layer by providing 'layer_names' (exclusive) or 'layer_indices'.

File: quantus/helpers/model/pytorch_model.py [hunk 7]
```diff
@@ -219,3 +219,67 @@ def sample(
...
+    ) -> np.ndarray:
+
+        device = self.device if self.device is not None else "cpu"
+        all_layers = [*self.model.named_modules()]
+        num_layers = len(all_layers)
+
+        if layer_indices is None:
+            layer_indices = []
+
+        # Convert negative indices to positive
```
Comment (annahedstroem) [inline quantus/helpers/model/pytorch_model.py:237]: Sorry, but for code cleanliness, all comments should be ended with a dot.

File: quantus/helpers/model/pytorch_model.py [hunk 8]
```diff
@@ -219,3 +219,67 @@ def sample(
...
+        positive_layer_indices = [
```
Comment (annahedstroem) [inline quantus/helpers/model/pytorch_model.py:238]: Please leave a comment what is meant with positive layer indices?

File: quantus/helpers/model/pytorch_model.py [hunk 9]
```diff
@@ -219,3 +219,67 @@ def sample(
...
+
+        if layer_names is None:
+            layer_names = []
+
+        def is_layer_of_interest(index, name):
+            if layer_names == [] and positive_layer_indices == []:
+                return True
+            return index in positive_layer_indices or name in layer_names
+
+        # skip modules defined by subclassing API
```
Comment (annahedstroem) [inline quantus/helpers/model/pytorch_model.py:250]: Formatting of comments... please check everywhere.

File: quantus/helpers/model/pytorch_model.py [hunk 10]
```diff
@@ -219,3 +219,67 @@ def sample(
...
+                    l[1], (self.model.__class__, torch.nn.Sequential)
+                ),
+                all_layers,
+            )
+        )
+
+        batch_size = x.shape[0]
+        hidden_outputs = []
+
+        # We register forward hook on layers of interest,
```
Comment (annahedstroem) [inline quantus/helpers/model/pytorch_model.py:263]: Fix

File: quantus/helpers/model/tf_model.py [hunk 11]
```diff
@@ -6,28 +6,49 @@
...
 
-from quantus.helpers import utils
 from quantus.helpers.model.model_interface import ModelInterface
+from quantus.helpers import utils
 
 
 class TensorFlowModel(ModelInterface):
     """Interface for tensorflow models."""
 
+    # All kwargs supported by Keras API
```
Comment (annahedstroem) [inline quantus/helpers/model/tf_model.py:29]: comment formatting

File: quantus/helpers/model/tf_model.py [hunk 12]
```diff
@@ -49,43 +70,50 @@ def __init__(
             softmax=softmax,
             model_predict_kwargs=model_predict_kwargs,
         )
+        self.cache = LRUCache(100)
```
Comment (annahedstroem) [inline quantus/helpers/model/tf_model.py:73]: Great can you add a comment here pls

File: quantus/helpers/model/tf_model.py [hunk 13]
```diff
@@ -49,43 +70,50 @@ def __init__(
...
-            Keyword arguments.
-
-        Returns
-        --------
-        np.ndarray
-            predictions of the same dimension and shape as the input, values in the range [0, 1].
-        """
-
+    def _get_predict_kwargs(self, **kwargs: Dict[str, ...]) -> Dict[str, ...]:
         # Use kwargs of predict call if specified, but don't overwrite object attribute
```
Comment (annahedstroem) [inline quantus/helpers/model/tf_model.py:76]: Please add numpydoc docstring

File: quantus/helpers/model/tf_model.py [hunk 14]
```diff
@@ -49,43 +70,50 @@ def __init__(
...
-
+        all_kwargs = {**self.model_predict_kwargs, **kwargs}
+        # Filter only ones which are supported by Keras
+        predict_kwargs = {
+            k: all_kwargs[k] for k in all_kwargs if k in self._available_predict_kwargs
+        }
+        return predict_kwargs
+
+    @cachedmethod(operator.attrgetter("cache"))
+    def _build_model_with_linear_top(self) -> Model:
```
Comment (annahedstroem) [inline quantus/helpers/model/tf_model.py:85]: Please add numpydoc docstring

File: quantus/helpers/model/tf_model.py [hunk 15]
```diff
@@ -49,43 +70,50 @@ def __init__(
...
+        # In this case model has a softmax on top, and we want linear.
+        # We have to rebuild the model and replace top with linear activation.
         config = self.model.layers[-1].get_config()
-        config["activation"] = target_act
-
+        config["activation"] = activations.linear
```
Comment (annahedstroem) [inline quantus/helpers/model/tf_model.py:89]: why specifying linear here?

File: quantus/helpers/model/tf_model.py [hunk 16]
```diff
@@ -49,43 +70,50 @@ def __init__(
...
         weights = self.model.layers[-1].get_weights()
 
         output_layer = Dense(**config)(self.model.layers[-2].output)
         new_model = Model(inputs=[self.model.input], outputs=[output_layer])
         new_model.layers[-1].set_weights(weights)
+        return new_model
+
+    def predict(self, x: np.ndarray, **kwargs) -> np.ndarray:
+        """Predict on the given input."""
```
Comment (annahedstroem) [inline quantus/helpers/model/tf_model.py:98]: Please add numpydoc docstring

File: quantus/helpers/model/tf_model.py [hunk 17]
```diff
@@ -49,43 +70,50 @@ def __init__(
...
+        # Generally, one should always prefer keras predict to __call__
+        # https://keras.io/getting_started/faq/#whats-the-difference-between-model-methods-predict-and-call
```
Comment (annahedstroem) [inline quantus/helpers/model/tf_model.py:100]: comments... fix everywhere! :)

File: quantus/helpers/model/tf_model.py [hunk 18]
```diff
@@ -49,43 +70,50 @@ def __init__(
...
+        predict_kwargs = self._get_predict_kwargs(**kwargs)
+
+        output_activation = self.model.layers[-1].activation
+        target_activation = activations.softmax if self.softmax else activations.linear
```
Comment (annahedstroem) [inline quantus/helpers/model/tf_model.py:104]: great.

File: quantus/helpers/model/tf_model.py [hunk 19]
```diff
@@ -95,24 +123,8 @@ def shape_input(
...
-            Indicates of the image dimensions are channel first, or channel last.
-            Inferred from the input shape if None.
-        batched: boolean
-            Indicates if the first dimension should be expanded or not, if it is just a single instance.
-
-        Returns
-        -------
-        np.ndarray
-            A reshaped input.
+        Reshape input into model-expected input.
```
Comment (annahedstroem) [inline quantus/helpers/model/tf_model.py:126]: What happened to this docstring

File: quantus/helpers/model/tf_model.py [hunk 20]
```diff
@@ -171,6 +171,71 @@ def get_random_layer_generator(self, order: str = "top_down", seed: int = 42):
             if order == "independent":
                 random_layer_model.set_weights(original_parameters)
             weights = layer.get_weights()
-            np.random.seed(seed=seed + 1)
+            np.random.seed(seed=seed + 1)  # noqa
```
Comment (annahedstroem) [inline quantus/helpers/model/tf_model.py:174]: why noqa?

File: quantus/helpers/model/tf_model.py [hunk 21]
```diff
@@ -171,6 +171,71 @@ def get_random_layer_generator(self, order: str = "top_down", seed: int = 42):
...
             layer.set_weights([np.random.permutation(w) for w in weights])
             yield layer.name, random_layer_model
+
+    @cachedmethod(operator.attrgetter("cache"))
+    def _build_hidden_representation_model(
+        self, layer_names: Tuple, layer_indices: Tuple
+    ) -> Model:
+        # Instead of rebuilding model on each image, which is evaluated by metric, we cache it
```
Comment (annahedstroem) [inline quantus/helpers/model/tf_model.py:182]: Please add numpydoc docstring

File: quantus/helpers/model/tf_model.py [hunk 22]
```diff
@@ -171,6 +171,71 @@ def get_random_layer_generator(self, order: str = "top_down", seed: int = 42):
...
+        return hidden_representation_model
+
+    def get_hidden_representations(
+        self,
+        x: np.ndarray,
+        layer_names: Optional[List[str]] = None,
+        layer_indices: Optional[List[int]] = None,
+        **kwargs,
+    ) -> np.ndarray:
+        # List is not hashable, so we pass names + indices as tuples
```
Comment (annahedstroem) [inline quantus/helpers/model/tf_model.py:210]: Please add numpydoc docstring

File: quantus/helpers/model/tf_model.py [hunk 23]
```diff
@@ -171,6 +171,71 @@ def get_random_layer_generator(self, order: str = "top_down", seed: int = 42):
...
+            layer_indices = []
+
+        # Convert negative indices to positive
+        positive_layer_indices = [
+            i if i >= 0 else num_layers + i for i in layer_indices
+        ]
+        if layer_names is None:
+            layer_names = []
+
+        hidden_representation_model = self._build_hidden_representation_model(
```
Comment (annahedstroem) [inline quantus/helpers/model/tf_model.py:224]: Please add comments in this function

File: quantus/helpers/model/tf_model.py [hunk 24]
```diff
@@ -171,6 +171,71 @@ def get_random_layer_generator(self, order: str = "top_down", seed: int = 42):
...
+            tuple(layer_names), tuple(positive_layer_indices)
+        )
+        predict_kwargs = self._get_predict_kwargs(**kwargs)
+        internal_representation = hidden_representation_model.predict(
+            x, **predict_kwargs
+        )
+        input_batch_size = x.shape[0]
+
+        if isinstance(internal_representation, np.ndarray):
+            # If we requested outputs only of 1 layer, keras will already return np.ndarray
```
Comment (annahedstroem) [inline quantus/helpers/model/tf_model.py:234]: Comment goes above if statement

**Review by annahedstroem — COMMENTED** [review]
Please make sure to re-run the API documentation creation. Will continue peer review tomorrow.

File: quantus/metrics/base_batched.py [hunk 25]
```diff
@@ -383,6 +384,33 @@ def evaluate_instance(
             "evaluate_instance() not implemented for BatchedMetric"
         )
 
+    def generate_normalized_explanations_batch(
```
Comment (annahedstroem) [inline quantus/metrics/base_batched.py:387]: generate_normalised_explanations_batch

File: requirements_test.txt [hunk 26]
```diff
@@ -17,3 +17,4 @@ torch>=1.10.1
 torchvision>=0.11.2
 zennit>=0.4.5 ; python_version > '3.6'
 tqdm>=4.62.3
+pytest-mock==3.8.2
```
Comment (annahedstroem) [inline requirements_test.txt:20]: What benefits do pytest-mock have?

**Review by annahedstroem — COMMENTED** [review]
:)

File: requirements_test.txt (hunk 26 above)
Comment (aaarrti) [reply requirements_test.txt:20]: `test_tf_model.test_predict` asserts against concrete values.
The test checks if `quantus.TensorFlowModel` wrapper correctly applies softmax or linear activations. However, the actual output values depend on the output of the underlying model, which is not part of Quantus codebase. So, I just mock the `tf.keras.predict`.
Afaik, it is common practice to use mocks, when doing asserts on component, which depend on 3-rd party components.

File: quantus/functions/perturb_func.py (hunk 3 above)
Comment (aaarrti) [reply quantus/functions/perturb_func.py]: fixed

File: quantus/functions/perturb_func.py (hunk 4 above)
Comment (aaarrti) [reply quantus/functions/perturb_func.py]: Honestly, I don't know. Changed to `low=-amplitude`

File: quantus/helpers/model/model_interface.py (hunk 5 above)
Comment (aaarrti) [reply quantus/helpers/model/model_interface.py]: fixed

File: quantus/metrics/base_batched.py (hunk 25 above)
Comment (aaarrti) [reply quantus/metrics/base_batched.py]: done

File: quantus/helpers/model/model_interface.py (hunk 6 above)
Comment (aaarrti) [reply quantus/helpers/model/model_interface.py]: done

File: quantus/helpers/model/tf_model.py (hunk 15 above)
Comment (aaarrti) [reply quantus/helpers/model/tf_model.py:89]: If we reach up this part, it means last layers `config["activation"] == activations.softmax`, and we must replace it by 
no activation (called linear in Keras)

File: quantus/helpers/model/tf_model.py (hunk 21 above)
Comment (aaarrti) [reply quantus/helpers/model/tf_model.py]: done

File: quantus/helpers/model/tf_model.py (hunk 22 above)
Comment (aaarrti) [reply quantus/helpers/model/tf_model.py]: Is the doc string in base class not sufficient?

File: quantus/helpers/model/tf_model.py (hunk 23 above)
Comment (aaarrti) [reply quantus/helpers/model/tf_model.py:238]: Is there a specific part which needs more clarification?

File: quantus/helpers/model/tf_model.py (hunk 24 above)
Comment (aaarrti) [reply quantus/helpers/model/tf_model.py]: done

File: quantus/helpers/model/tf_model.py (hunk 20 above)
Comment (aaarrti) [reply quantus/helpers/model/tf_model.py]: I was annoyed by this warning 
image
Anyway, removed.

File: quantus/helpers/model/pytorch_model.py (hunk 8 above)
Comment (aaarrti) [reply quantus/helpers/model/pytorch_model.py:239]: done

File: quantus/helpers/model/tf_model.py (hunk 12 above)
Comment (aaarrti) [reply quantus/helpers/model/tf_model.py:76]: done

File: quantus/helpers/model/tf_model.py (hunk 13 above)
Comment (aaarrti) [reply quantus/helpers/model/tf_model.py]: done

File: quantus/helpers/model/tf_model.py (hunk 14 above)
Comment (aaarrti) [reply quantus/helpers/model/tf_model.py:90]: done

File: quantus/helpers/model/tf_model.py (hunk 22 above)
Comment (aaarrti) [reply quantus/helpers/model/tf_model.py]: I checked, PyCharm renders docs from the doc string of the base class.

File: requirements_test.txt (hunk 26 above)
Comment (annahedstroem) [reply requirements_test.txt:20]: Nice. Thanks.

File: quantus/helpers/model/tf_model.py (hunk 22 above)
Comment (annahedstroem) [reply quantus/helpers/model/tf_model.py]: You mean for the readthedocs?

Comment (aaarrti) [reply quantus/helpers/model/tf_model.py]: done

File: quantus/helpers/model/pytorch_model.py (hunk 7 above)
Comment (aaarrti) [reply quantus/helpers/model/pytorch_model.py]: done

File: quantus/helpers/model/pytorch_model.py (hunk 9 above)
Comment (aaarrti) [reply quantus/helpers/model/pytorch_model.py]: done

File: quantus/helpers/model/tf_model.py (hunk 11 above)
Comment (aaarrti) [reply quantus/helpers/model/tf_model.py]: done

File: quantus/helpers/model/pytorch_model.py (hunk 10 above)
Comment (aaarrti) [reply quantus/helpers/model/pytorch_model.py]: done

File: quantus/helpers/model/tf_model.py (hunk 17 above)
Comment (aaarrti) [reply quantus/helpers/model/tf_model.py]: done

File: quantus/helpers/model/tf_model.py (hunk 19 above)
Comment (aaarrti) [reply quantus/helpers/model/tf_model.py:135]: fixed

File: quantus/helpers/model/tf_model.py (hunk 13 above)
Comment (aaarrti) [reply quantus/helpers/model/tf_model.py]: done

**Comment by aaarrti** [comment]
This pull request is too big. There are many changes, which are out of the scope. 
I will split it up in a few smaller ones, which are easier to review.

**Pull request closed without merging**